src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/W3CPropagator.cs

```diff
@@ -0,0 +1,686 @@
+                return false;
+            }
+
+            value = (byte)(((uint)byte1 << 4) + byte2);
+            return true;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool TryDecodeHexDigit(char c, out byte value)
+        {
+            value = (byte)HexConverter.FromChar((int)c);
+            return value != 0xFF; // invalid hex digit
+        }
+
+        // Allowed baggage key characters:
+        //  tchar = "!" / "#" / "$" / "%" / "&" / "'" / "*" / "+" / "-" / "." / "^" / "_" / "`" / "|" / "~" / DIGIT / ALPHA
+        //  DIGIT = 0-9
+        //  ALPHA = A-Z / a-z
+#if NET
+        private const string BaggageKeyValidCharacters = "!#$%&'*+-.0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ^_`abcdefghijklmnopqrstuvwxyz|~";
+        private static readonly SearchValues<char> s_validBaggageKeyChars = SearchValues.Create(BaggageKeyValidCharacters);
+
+        private static bool IsInvalidBaggageKey(ReadOnlySpan<char> span) => span.ContainsAnyExcept(s_validBaggageKeyChars);
+#else
+        private static ulong[] s_invalidBaggageKeyCharsMask = [0xFC009305FFFFFFFF, 0x2800000038000001];
+
+        private static bool IsInvalidBaggageKey(ReadOnlySpan<char> span)
+        {
+            foreach (char c in span)
+            {
+                // key support only ascii characters according to the W3C specs
+                if (c >= 0x07F || ((s_invalidBaggageKeyCharsMask[c >> 6] & (ulong)((ulong)1 << ((int)c & 63))) != 0))
+                {
+                    return true;
+                }
+            }
+
+            return false; // valid key
+        }
+#endif
+
+        // key = ( lcalpha / DIGIT ) 0*255 ( keychar )
+        // keychar    = lcalpha / DIGIT / "_" / "-"/ "*" / "/" / "@"
+        // lcalpha    = %x61-7A ; a-z
+
+#if NET
+        private const string TraceStateKeyValidChars = "*-/@_abcdefghijklmnopqrstuvwxyz";
+        private static readonly SearchValues<char> s_validTraceStateChars = SearchValues.Create(TraceStateKeyValidChars);
+
+        private static bool IsInvalidTraceStateKey(ReadOnlySpan<char> span) => span.ContainsAnyExcept(s_validTraceStateChars);
+
+        private const string TraceStateValueValidChars = "!\"#$%&'()*+-./0123456789:;<>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~";
+        private static readonly SearchValues<char> s_validTraceStateValueChars = SearchValues.Create(TraceStateValueValidChars);
+
+        private static bool IsInvalidTraceStateValue(ReadOnlySpan<char> span) => span.ContainsAnyExcept(s_validTraceStateValueChars);
+#else
+        private static ulong[] ValidTraceStateKeyCharsMask = [0x0000A40000000000, 0x07FFFFFE80000001];
+
+        private static bool IsInvalidTraceStateKey(ReadOnlySpan<char> key)
+        {
+            if (key.IsEmpty || (key[0] < 'a' || key[0] > 'z')) // Key has to start with a lowercase letter
+            {
+                return true; // invalid key character, skip current entry
+            }
```

Comment:
This condition isn't matched in the `#if` above.
Same for `IsEmpty` check in `IsInvalidTraceStateValue` below.

Reply:
Ooops. Thanks for the catch.